Shift from Travis CI to Github Actions

ankushbhardwxj opened the pull request:
This PR aims to shift the mason registry from the existing Travis CI to Github Actions. The packages submitted by an user would be tested and built using workflows and actions. There are 2 modes of workflows : - **pre-merge** workflow & **post-merge** workflow. 

- pre-merge workflow ( `masonRegCI.yaml` ) : This is triggered when a PR is made on the repository. This workflow is responsible for checking if a TOML file exists inside the packages in the Bricks directory. Additionally, Chapel is built and also a CI check is done in the package using `mason publish --ci-check`. 
- post-merge workflow ( `genCacheOnPush.yaml` ) : This is triggered once a PR is merged and a push is made on repository. This workflow is responsible for computing a score for the package based on quality and update the `cache.toml` file on repository. If a package with a certain version is already present in cache, then this script is aborted and cache is not updated. 

Additional post-merge changes to make : 
- [ ] change credentials to that of registry
- [ ] update cache file with computed values of already present packages
- [ ] improve `README `file to reflect new changes made and present guidelines for users.

ankushbhardwxj committed:
new scripts & shift to Github Actions

ankushbhardwxj commented:
@ben-albrecht Could you provide me the credentials for `git config --global user.name` and `git config --global user.email` for this repository ? 
Also, should I calculate scores for existing packages and update the cache ? (without testing them since they fail `mason publish --ci-check`)
Another small question, would you like the checks for git tag versioning, namespace collisions, be done in our bash scripts or in `mason publish --ci-check` ?

ben-albrecht commented:
> Also, should I calculate scores for existing packages and update the cache ? (without testing them since they fail mason publish --ci-check)

Do you mean compute the scores then manually add them? This sounds reasonable to me.

> Another small question, would you like the checks for git tag versioning, namespace collisions, be done in our bash scripts or in mason publish --ci-check ?

I would prefer `mason publish` since that gives devs an opportunity to try this locally and catch these errors before opening a PR.

ben-albrecht commented on .github/workflows/genCacheOnPush.yaml:
```diff
@@ -0,0 +1,40 @@
+      - name: Commit changes
+        run: |
+          git config --global user.name "ankingcodes"
+          git config --global user.name "ankush.bhardwaj0@gmail.com"
```
Let's put dummy place holders here until we get chapelautomaton credentials for this.

ben-albrecht commented on .github/workflows/genCacheOnPush.yaml:
```diff
@@ -0,0 +1,40 @@
+    paths-ignore:
+      - '.github/**'
+      - 'util/**'
+jobs:
```
Is there an easy way to disable this script for now, but still check it in?

ben-albrecht commented on util/searchRankScript.bash:
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env bash
+# Clone the package to be published from latest commit
```
Just checking - did you end up running `shellcheck` on this?

ankushbhardwxj replied:
Yeah, I have already resolved shell check errors.

ben-albrecht commented on util/searchRankScript.bash:
```diff
@@ -0,0 +1,68 @@
+FILE=$package
+basename "$FILE"
+f="$(basename -- "$FILE")"
+source="$(grep source "$f"| cut -d= -f2)"
```
I don't think it'd hurt to emit verbose log messages in this file, in case something goes wrong due to user error, e.g. we could print these grepped / translated variables as we parse them.

ben-albrecht replied:
If you want to differentiate useful checkpoints vs. verbose logging, you can add some prefix like:

```sh
echo "[LOG]: name = ${name}"
```

ben-albrecht commented on .github/workflows/genCacheOnPush.yaml:
```diff
@@ -0,0 +1,40 @@
+          python-version: "3.7"
+      - name: Sort cache file
+        run: |
+          pip install toml-sort
```
We should pin (at least a major) version of toml-sort here. Using a `requirements.txt` file would be even better.

ankushbhardwxj replied:
Do you mean you want to do a `pip install toml-sort==0.17.0` here ?

ben-albrecht replied:
That works for me!

ankushbhardwxj replied:
@ben-albrecht turns out the author of `toml-sort` hadn't configured the package properly, so we can only get the latest version of toml-sort, i.e., 0.18.0, if we use `pip install toml-sort==0.17.1` we get an error message as follows: 
```
 Could not find a version that satisfies the requirement toml-sort==0.17.1 (from versions: )
No matching distribution found for toml-sort==0.17.1
```

ben-albrecht replied:
Can we use `toml-sort==0.18.0`?

ankushbhardwxj replied:
Unfortunately, no. As you can see from the above error message that the versions array in package configuration is empty.

ben-albrecht commented on .github/workflows/genCacheOnPush.yaml:
```diff
@@ -0,0 +1,40 @@
+          git config --global user.name "ankingcodes"
+          git config --global user.name "ankush.bhardwaj0@gmail.com"
+          git add cache.toml
+          git commit -m "appended to cache.toml"
```
Is it possible to embed a reference to the PR or commit that triggered this job into the commit message?

e.g.

```
git commit -m "append package from PR #nnn to cache.toml"
# or
git commit -m "append package from merge commit <SHA> to cache.toml"
```

This would be very helpful in debugging if something ever goes wrong.

ben-albrecht commented on util/searchRankScript.bash:
```diff
@@ -0,0 +1,68 @@
+  exit 0
+fi
+path="$package$end"
+cd "$(dirname "$path")" || exit 1
```
I generally prefer not changing the directory (especially if there are multiple changes) in scripts like these, because it introduces a "state" to our script which can make debugging / reproducing issues more difficult.

I'd suggest prepending paths where in the variables used and staying in the top-level directory if nothing prohibits that.

ankushbhardwxj committed:
refactored scripts and disabled onpush workflow

ankushbhardwxj commented:
@ben-albrecht Made requested changes. 
- disabled on-push script `genCacheOnPush.yaml` by renaming it to `genCacheOnPush.yaml.disabled`, since only yaml files are run in Github Actions runner environment.
- refactored `searchRankScript.bash` to perform operations from root. Also, added verbose messages for easy debugging as follows: 
```
[LOG]: Package path from git command = Bricks/test/0.1.0.toml
[LOG]: path = Bricks/test [LOG]: Package TOML file = 0.1.0.toml 
[LOG]: Package url = https://gitlab.com/npadmana/numpylike.git 
[LOG]: Package name = NumpyLike 
[LOG]: Version = 0.1.0 Cloning into 'Bricks/test/newPackage'... 
remote: Enumerating objects: 14, done. 
remote: Counting objects: 100% (14/14), done. 
remote: Compressing objects: 100% (10/10), done. 
remote: Total 14 (delta 1), reused 0 (delta 0), pack-reused 0 
Unpacking objects: 100% (14/14), done. 
NumpyLike cloned succesfully at Bricks/test/newPackage 
ModuleFile found in package 
Mason.toml found in package 
test dir found in package 
src dir found in package 
Found 0 examples in package. 
Found 1 tests in package. 
Total score generated for NumpyLike = 5 
Wrote 'NumpyLike = 5' to cache.toml 
```
- Improved onpush script to use latest commit SHA in the next commit message
